examples/cms/src/FieldMapping.tsx

```diff
@@ -0,0 +1,225 @@
+import { useState, useMemo } from "react"
+import { Fragment } from "react/jsx-runtime"
+import { DataSource, FieldConfig, computeFieldConfigs } from "./data"
+import { isNotNull } from "./utils"
+
+interface FieldMappingRowProps {
+    fieldConfig: FieldConfig
+    isIgnored: boolean
+    onFieldToggle: (fieldId: string) => void
+    onFieldNameChange: (fieldId: string, name: string) => void
+}
+
+function FieldMappingRow({ fieldConfig, isIgnored, onFieldToggle, onFieldNameChange }: FieldMappingRowProps) {
+    const field = fieldConfig.field
+
+    const isUnsupported = !field
+    const hasFieldNameChanged = field!.name !== fieldConfig.name
+    const fieldName = hasFieldNameChanged ? field!.name : ""
+    const placeholder = isUnsupported ? "Unsupported Field" : fieldConfig.name
+
+    const isDisabled = useMemo(() => isUnsupported || isIgnored, [isUnsupported, isIgnored])
+
+    return (
+        <Fragment>
+            <div
+                className="source-field"
+                aria-disabled={isDisabled}
+                onClick={() => onFieldToggle(fieldConfig.field!.id)}
+                role="button"
+            >
+                <input
+                    type="checkbox"
+                    disabled={isUnsupported}
+                    checked={!isDisabled}
+                    onChange={e => {
+                        e.stopPropagation()
+                    }}
+                />
+                <span>{fieldConfig.name}</span>
+            </div>
+            <svg xmlns="http://www.w3.org/2000/svg" width="8" height="16">
+                <path
+                    d="M 3 11 L 6 8 L 3 5"
+                    fill="transparent"
+                    strokeWidth="1.5"
+                    stroke="#999"
+                    strokeLinecap="round"
+                    strokeLinejoin="round"
+                ></path>
+            </svg>
+            <input
+                type="text"
+                style={{
+                    width: "100%",
+                    opacity: isDisabled ? 0.5 : 1,
+                }}
+                disabled={isDisabled}
+                placeholder={placeholder}
+                value={fieldName || ""}
+                onChange={e => {
+                    if (!field) return
+                    if (!e.target.value.trim()) {
+                        onFieldNameChange(field.id, fieldConfig.name)
+                    } else {
+                        onFieldNameChange(field.id, e.target.value)
+                    }
+                }}
+            />
+        </Fragment>
+    )
+}
+
+export function FieldMapping({
+    savedFieldsConfig,
+    existingFields,
+    dataSource,
+    savedSlugFieldId,
+    onSubmit,
+}: {
+    savedFieldsConfig: FieldConfig[] | undefined
+    existingFields: ManagedCollectionField[]
+    dataSource: DataSource
+    savedSlugFieldId: string | null
+    onSubmit: (dataSource: DataSource, fields: FieldConfig[], slugFieldId: string) => Promise<void>
+}) {
```

Comment:
I mostly prefer if component props are extracted into a standalone interface. Do you prefer inline like this?